README.md
# PHP-API-CLIENT
Cliente PHP para consumir el API de Flow.

**Flow.cl** es una pasarela de pagos para comercio electrónico. Este cliente le permite integrar su ecommerce para recibir pagos online.

## Requerimientos
* PHP 5.2 o superior
* php_curl activado
* php_openssl activado

## Instalación
Como ejemplo del uso de esta API-REST, baje la última versión de https://github.com/flowcl/PHP-API-CLIENT y copie los archivos en su servidor.

## Documentación
La documentación completa del API REST de Flow la encuentrá aquí: https://www.flow.cl/docs/api.html

## Comenzando
### Configurando el cliente
Configure correctamente el cliente en el archivo **lib/Config.class.php**.
Lo primero que debe configurar es su apiKey, secretKey. Para ello, acceda a https://www.flow.cl, una vez autenticado con su cuenta Flow, seleccione **Mis datos**  y recupere su apiKey y secretKey desde la sección Seguridad.

| Sitio | Mi cuenta flow 	|
|---	|---	|
| Production  	| https://www.flow.cl/app/web/misDatos.php   	|
| Sandbox  	| https://sandbox.flow.cl/app/web/misDatos.php  	|



- **APIKEY** el apiKey obtenida desde su cuenta Flow
- **SECRETKEY** el secretKey obtenida desde su cuenta Flow
- **APIURL** la URL del endpoint del API de Flow, Aquí podrá configurar el endpoint de producción o del sandbox. 
- **BASEURL** La URL base donde instaló el cliente PHP en su servidor

La URL del endpoint del API Flow, puede ser:

| Sitio | mi cuenta flow 	|
|---	|---	|
| Production  	| https://www.flow.cl/api   	|
| Sandbox  	| https://sandbox.flow.cl/api 	|

información basada en documentación: https://www.flow.cl/docs/api.html#section/Introduccion/Acceso-al-API



```php
$commerceConfig = array(
 	"APIKEY" => "1F90971E-8276-4713-97FF-2BLF5091EE3B", // Registre aquí su apiKey
 	"SECRETKEY" => "f8b45f9b8bcdb5702dc86a1b894492303741c405", // Registre aquí su secretKey
 	"APIURL" => "https://www.flow.cl/api", // Producción EndPoint o Sandbox EndPoint
 	"BASEURL" => "https://www.misitio/apiFlow" //Registre aquí la URL base en su página donde instaló el cliente
 );
```
### Llamando a un servicio
En este ejemplo crearemos una Orden de Cobro y redireccionaremos el browser del pagador para efectuar el pago
```php
<?php
/**
 * Ejemplo de creación de una orden de cobro, iniciando una transacción de pago
 * Utiliza el método payment/create
 */
require(__DIR__ . "/../../lib/FlowApi.class.php");

//Para datos opcionales campo "optional" prepara un arreglo JSON
$optional = array(
	"rut" => "9999999-9",
	"otroDato" => "otroDato"
);
$optional = json_encode($optional);

//Prepara el arreglo de datos
$params = array(
	"commerceOrder" => random_int(1100,2000),
	"subject" => "Pago de prueba",
	"currency" => "CLP",
	"amount" => 5000,
	"email" => "cliente@gmail.com",
	"paymentMethod" => 9,
	"urlConfirmation" => Config::get("BASEURL") . "/examples/payments/confirm.php",
	"urlReturn" => Config::get("BASEURL") ."/examples/payments/result.php",
	"optional" => $optional
);
//Define el metodo a usar
$serviceName = "payment/create";

try {
	// Instancia la clase FlowApi
	$flowApi = new FlowApi;
	// Ejecuta el servicio
	$response = $flowApi->send($serviceName, $params,"POST");
	//Prepara url para redireccionar el browser del pagador
	$redirect = $response["url"] . "?token=" . $response["token"];
	header("location:$redirect");
} catch (Exception $e) {
	echo $e->getCode() . " - " . $e->getMessage();
}
?>
```
Otros ejemplos los podrá ver en la carpeta **examples** de este cliente.


CHANGELOG
f338b7b Fri Nov 1 12:12:40 2019 -0300 Merge pull request #6 from appwebd/general
a54e920 Fri Nov 1 12:10:21 2019 -0300 Error handling, PHPDoc.
ca9250f Fri Nov 1 11:49:49 2019 -0300 Merge pull request #5 from appwebd/bug_security_hotspot
341f5e3 Fri Nov 1 11:46:57 2019 -0300 security hotspot CVE-2006-3419 (Using pseudorandom number generators (PRNGs) is security-sensitive)
d3ac405 Tue Oct 29 23:30:32 2019 -0300 Merge pull request #4 from appwebd/bug_type_compatibility
a6e3482 Tue Oct 29 23:29:04 2019 -0300 type compatibility : https://www.php.net/manual/en/language.types.string.php
5cb9b6e Tue Oct 29 23:01:09 2019 -0300 Merge pull request #3 from appwebd/PHPDoc
ae7d038 Tue Oct 29 22:57:57 2019 -0300 PHPDoc for non-existing argument,  Return type does not march
dda933a Tue Oct 29 21:10:57 2019 -0300 was added Missing throws tags
023bd3f Tue Oct 29 21:07:53 2019 -0300 was added Missing throws tags
b87b056 Tue Oct 29 20:58:06 2019 -0300 Merge pull request #2 from appwebd/bug_code_style
2991534 Tue Oct 29 20:53:04 2019 -0300 fixed redundant closing tag
d727818 Tue Oct 29 20:33:47 2019 -0300 Merge pull request #1 from appwebd/caracteristica_unificar_mensajes
fbb621d Tue Oct 29 19:59:44 2019 -0300 Unification of messages to English (they should be to spanish in production platform?).
de5e321 Tue Oct 29 19:50:10 2019 -0300 was added Change Log
af34b4d Tue Oct 29 19:26:53 2019 -0300 Initial commit
1b79039 Fri Oct 11 14:19:47 2019 -0300 Update FlowApi.class.php
36810f9 Fri Jul 26 13:13:43 2019 -0400 Revert "Update FlowApi.class.php"
8738bb6 Fri Jul 26 12:51:17 2019 -0400 Merge branch 'master' of https://github.com/flowcl/PHP-API-CLIENT
ace67f3 Fri Jul 26 12:49:13 2019 -0400 Update FlowApi.class.php
bbbcf0a Thu Jul 26 18:44:06 2018 -0400 Update Read.me
620fe5e Thu Jul 26 15:05:40 2018 -0400 Update Read.me
76dc5b1 Thu Jul 26 15:04:53 2018 -0400 Update Read.me
45debb3 Thu Jul 26 15:02:17 2018 -0400 Actualiza las url
84e45e5 Thu Jul 26 14:49:27 2018 -0400 Merge branch 'master' of https://github.com/flowcl/PHP-API-CLIENT
f9e9414 Thu Jul 26 14:49:21 2018 -0400 Update payments/create
605d20f Thu Jul 26 14:34:55 2018 -0400 Update Read.me
aa46125 Thu Jul 26 13:21:02 2018 -0400 Update Read.me
2a22902 Thu Jul 26 13:14:31 2018 -0400 Update Read.me
da44095 Thu Jul 26 13:08:04 2018 -0400 Agrega información a Read.me
211cd15 Thu Jul 26 12:15:47 2018 -0400 modifica require
f88edbc Thu Jul 26 12:13:39 2018 -0400 Agrega ejemplos payments
5fa37c4 Thu Jul 26 11:08:54 2018 -0400 Agrega componentes lib
0451aa7 Thu Jul 26 09:58:38 2018 -0400 Initial commit


lib/Config.class.php
<?php
/**
 * Clase para Configurar el cliente
 * PHP Version 5.2
 *
 * @Filename: Config.class.php
 * @version: 2.0
 * @Author: flow.cl
 * @Email: csepulveda@tuxpan.com
 * @Date: 28-04-2017 11:32
 * @Last Modified by: Carlos Sepulveda
 * @Last Modified time: 28-04-2017 11:32
 */

$COMMERCE_CONFIG = array(
    "APIKEY" => "mi apiKey", // Registre aquí su apiKey
    "SECRETKEY" => "mi secretKey", // Registre aquí su secretKey
    "APIURL" => "https://www.flow.cl/api", // Producción EndPoint o Sandbox EndPoint
    "BASEURL" => "https://www.micomercio.cl/apiFlow" //Registre aquí la URL base en su página donde instalará el cliente
);

class Config
{
    /**
     * Get item of structure $commerce_config
     *
     * @param string $name
     *
     * @return mixed
     * @throws Exception
     */
    static function get($name)
    {
        global $COMMERCE_CONFIG;
        if (!isset($COMMERCE_CONFIG[$name])) {
            throw new Exception("The configuration element does not exist", 1);
        }
        return $COMMERCE_CONFIG[$name];
    }
}


lib/FlowApi.class.php
<?php

/**
 * Clase cliente del Api2 de Flow
 * PHP Version 5.2
 *
 * @Filename: FlowApi.class.php
 * @version: 2.1
 * @Author: flow.cl
 * @Email: csepulveda@tuxpan.com
 * @Date: 28-04-2017 11:32
 * @Last Modified by: Carlos Sepulveda
 * @Last Modified time: 28-04-2017 11:32
 */

require(__DIR__ . "/Config.class.php");

class FlowApi
{

    protected $apiKey;
    protected $secretKey;


    public function __construct()
    {

        try {
            $this->apiKey = Config::get("APIKEY");
            $this->secretKey = Config::get("SECRETKEY");
        } catch (Exception $e) {
            echo $e->getCode(), ' ', $e->getMessage();
        }

    }


    /**
     * Funcion que invoca un servicio del Api de Flow
     *
     * @param string $service Nombre del servicio a ser invocado
     * @param array $params datos a ser enviados
     * @param string $method metodo http a utilizar
     *
     * @return string en formato JSON
     * @throws Exception
     */
    public function send($service, $params, $method = "GET")
    {
        $method = strtoupper($method);
        $url = Config::get("APIURL") . "/" . $service;
        $params = array("apiKey" => $this->apiKey) + $params;
        $params["s"] = $this->sign($params);
        if ($method == "GET") {
            $response = $this->httpGet($url, $params);
        } else {
            $response = $this->httpPost($url, $params);
        }

        if (isset($response["info"])) {
            $code = $response["info"]["http_code"];
            if (!in_array($code, array("200", "400", "401"))) {
                throw new Exception("An unexpected HTTP_CODE error occurred: " . $code, $code);
            }
        }
        $body = json_decode($response["output"], true);
        return $body;
    }

    /**
     * Funcion que firma los parametros
     *
     * @param array $params Parametros a firmar
     *
     * @return string de firma
     * @throws Exception
     */
    private function sign($params)
    {
        $keys = array_keys($params);
        sort($keys);
        $toSign = "";
        foreach ($keys as $key) {
            $toSign .= $key . $params[$key];
        }
        if (!function_exists("hash_hmac")) {
            throw new Exception("hash_hmac function does not exist", 1);
        }
        return hash_hmac('sha256', $toSign, $this->secretKey);
    }

    /**
     * Funcion que hace el llamado via http GET
     *
     * @param string $url url a invocar
     * @param $params
     * @return array en formato JSON
     * @throws Exception
     */
    private function httpGet($url, $params)
    {
        $url = $url . "?" . http_build_query($params);
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
        $output = curl_exec($ch);
        if ($output === false) {
            $error = curl_error($ch);
            throw new Exception($error, 1);
        }
        $info = curl_getinfo($ch);
        curl_close($ch);
        return array("output" => $output, "info" => $info);
    }

    /**
     * Funcion que hace el llamado via http POST
     *
     * @param string $url url a invocar
     * @param array $params datos a enviar
     * @return array en formato JSON
     * @throws Exception
     */
    private function httpPost($url, $params)
    {
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
        curl_setopt($ch, CURLOPT_POST, TRUE);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $params);
        $output = curl_exec($ch);
        if ($output === false) {
            $error = curl_error($ch);
            throw new Exception($error, 1);
        }
        $info = curl_getinfo($ch);
        curl_close($ch);
        return array("output" => $output, "info" => $info);
    }

    /**
     * Function to set the apiKey and secretKey and not use those of the
     * configuration. Get your apiKey and secretKey depending on whether it is
     * in production or sandbox. (Both require registering separately).
     * https://www.flow.cl/api
     *
     * @param string $apiKey Api Key
     * @param string $secretKey Secret Key
     */
    public function setKeys($apiKey, $secretKey)
    {
        $this->apiKey = $apiKey;
        $this->secretKey = $secretKey;
    }


}


examples/payments/result.php
<?php
/**
 * Pagina del comercio para redireccion del pagador
 * A esta página Flow redirecciona al pagador pasando vía POST
 * el token de la transacción. En esta página el comercio puede
 * mostrar su propio comprobante de pago
 * PHP Version 5.2
 *
 * @category  View
 * @package   Result
 * @author    flow.cl
 * @copyright 2019 (c) copyright
 * @license   Default copyright laws apply
 * @version   GIT: <git_id>
 * @link      https://www.flow.cl
 * @date      29/10/2019 22:27
 */
require(__DIR__ . "/../../lib/FlowApi.class.php");

try {
    //Recibe el token enviado por Flow
    if (!isset($_POST["token"])) {
        throw new Exception('Token expected from flow.cl platform', 1);
    }
    $token = filter_input(INPUT_POST, 'token');
    $params = array(
        "token" => $token
    );
    //Indica el servicio a utilizar
    $serviceName = "payment/getStatus";
    $flowApi = new FlowApi();
    $response = $flowApi->send($serviceName, $params, "GET");

    print_r($response);

} catch (Exception $e) {
    echo 'Error: ' , $e->getCode() , ' - ' , $e->getMessage();
}



examples/payments/confirm.php
<?php
/**
 * Pagina del comercio para recibir la confirmación del pago
 * Flow notifica al comercio del pago efectuado
 * PHP Version 5.2
 *
 * @category  View
 * @package   confirm
 * @author    flow.cl
 * @copyright 2019 (c) copyright
 * @license   Default copyright laws apply
 * @version   GIT: <git_id>
 * @link      https://www.flow.cl
 * @date      29/10/2019 22:56
 */
require(__DIR__ . "/../../lib/FlowApi.class.php");

try {
    if (!isset($_POST["token"])) {
        throw new Exception("Token expected from flow.cl platform", 1);
    }
    $token = filter_input(INPUT_POST, 'token');
    $params = array(
        "token" => $token
    );
    $serviceName = "payment/getStatus";
    $flowApi = new FlowApi();
    $response = $flowApi->send($serviceName, $params, "GET");

    //Actualiza los datos en su sistema

    print_r($response);


} catch (Exception $e) {
    echo 'Error: ' , $e->getCode() , ' - ' , $e->getMessage();
}


examples/payments/createEmail.php
<?php
/**
 * Ejemplo de creación de una orden de cobro vái cobro por email
 * Utiliza el método payment/createEmail
 * PHP Version 5.2
 *
 * @category  View
 * @package   createEmail
 * @author    flow.cl
 * @copyright 2019 (c) copyright
 * @license   Default copyright laws apply
 * @version   GIT: <git_id>
 * @link      https://www.flow.cl
 * @date      29/10/2019 22:54
 */


require(__DIR__ . "/../../lib/FlowApi.class.php");



try {

    //Prepara los parámetros
    $params = array(
        "commerceOrder" => random_int(1100, 2000),
        "subject" => "pago prueba cobro Email",
        "currency" => "CLP",
        "amount" => 2000,
        "email" => "cliente@gmail.com",
        "paymentMethod" => 9,
        "urlConfirmation" => Config::get("BASEURL") . "/examples/payments/confirm.php",
        "urlReturn" => Config::get("BASEURL") . "/examples/payments/result.php",
        "forward_days_after" => 1,
        "forward_times" => 2
    );

    $serviceName = "payment/createEmail";

    $flowApi = new FlowApi;
    $response = $flowApi->send($serviceName, $params, "POST");

    print_r($response);


} catch (Exception $e) {
    echo 'Error: ' , $e->getCode() , ' - ' , $e->getMessage();
}



examples/payments/create.php
<?php
/**
 * Ejemplo de creación de una orden de cobro, iniciando una transacción de pago
 * Utiliza el método payment/create
 * PHP Version 5.2
 *
 * @category  View
 * @package   Create
 * @author    flow.cl
 * @copyright 2019 (c) copyright
 * @license   Default copyright laws apply
 * @version   GIT: <git_id>
 * @link      https://www.flow.cl
 * @date      29/10/2019 22:55
 */
require(__DIR__ . "/../../lib/FlowApi.class.php");


try {
    //Para datos opcionales campo "optional" prepara un arreglo JSON
    $optional = array(
        "rut" => "9999999-9",
        "otroDato" => "otroDato"
    );
    $optional = json_encode($optional);

    //Prepara el arreglo de datos
    $params = array(
        "commerceOrder" => random_int(1100, 2000),
        "subject" => "Pago de prueba",
        "currency" => "CLP",
        "amount" => 5000,
        "email" => "cliente@gmail.com",
        "paymentMethod" => 9,
        "urlConfirmation" => Config::get("BASEURL") . "/examples/payments/confirm.php",
        "urlReturn" => Config::get("BASEURL") . "/examples/payments/result.php",
        "optional" => $optional
    );
    //Define el metodo a usar
    $serviceName = "payment/create";


    // Instancia la clase FlowApi
    $flowApi = new FlowApi;
    // Ejecuta el servicio
    $response = $flowApi->send($serviceName, $params, "POST");
    //Prepara url para redireccionar el browser del pagador
    $redirect = $response["url"] . "?token=" . $response["token"];
    header("location:$redirect");
} catch (Exception $e) {
    echo 'Error: ' , $e->getCode() , ' - ' , $e->getMessage();
}

